### PR TITLE
feat(schedule): impl event_handler

### LINKED ISSUE #27: Send events through heartbeat response to CeresDB node
**Description**

Send events through heartbeat response to CeresDB node,  CeresDB node can change it's state by those events.

**Proposal**

One sub task of #10 

**Additional context**

### PR DESCRIPTION
**Pull request by chunshao90**
# Which issue does this PR close?

Closes #27 

# Rationale for this change

# What changes are included in this PR?
Send events through heartbeat response to CeresDB node, CeresDB node can change it's state by those events.
# Are there any user-facing changes?
none

# How does this change test

TODO: need to add tests

### CODE REVIEW

File: server/cluster/coordinator.go [hunk 1]
```diff
@@ -5,25 +5,65 @@ package cluster
...
 type coordinator struct {
 	// Mutex is to ensure only one coordinator can run at the same time.
-	sync.Mutex
+	lock    sync.Mutex
 	cluster *Cluster
+
+	ctx          context.Context
+	cancel       context.CancelFunc
+	bgJobWg      *sync.WaitGroup
+	eventHandler *schedule.EventHandler
```
Comment (ShiKaiWi) [inline server/cluster/coordinator.go:29]: `eventH` seems better.

File: server/schedule/event_handler.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+)
+
+type Event interface {
+	TransformToHeartbeatResp() *metaservicepb.NodeHeartbeatResponse
+}
+
+type NoneEvent struct{}
+
+func (*NoneEvent) TransformToHeartbeatResp() *metaservicepb.NodeHeartbeatResponse {
+	// TODO: add response header
```
Comment (ShiKaiWi) [inline server/schedule/event_handler.go:20]: Why not add the header here?

File: server/schedule/event_handler.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/pkg/errors"
+)
+
+type Event interface {
+	TransformToHeartbeatResp() *metaservicepb.NodeHeartbeatResponse
+}
+
+type NoneEvent struct{}
+
+func (*NoneEvent) TransformToHeartbeatResp() *metaservicepb.NodeHeartbeatResponse {
```
Comment (ShiKaiWi) [inline server/schedule/event_handler.go:19]: `ToHeartbeatResp` seems better.

File: server/schedule/event_handler.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	// TODO: add response header
+	return &metaservicepb.NodeHeartbeatResponse{Timestamp: uint64(time.Now().UnixMilli())}
+}
+
+type OpenEvent struct {
+	ShardIDs []uint32
+}
+
+func (open *OpenEvent) TransformToHeartbeatResp() *metaservicepb.NodeHeartbeatResponse {
```
Comment (ShiKaiWi) [inline server/schedule/event_handler.go:28]: `ToHeartbeatResp` seems better.

File: server/cluster/cluster.go [hunk 5]
```diff
@@ -43,6 +47,10 @@ func (c *Cluster) Name() string {
 func (c *Cluster) Load(ctx context.Context) error {
 	c.lock.Lock()
 	defer c.lock.Unlock()
+	if c.coordinator == nil {
```
Comment (ShiKaiWi) [inline server/cluster/cluster.go:50]: I guess the method name should be renamed to `Init` rather than `Load`.

File: server/cluster/cluster.go [hunk 6]
```diff
@@ -43,6 +47,10 @@ func (c *Cluster) Name() string {
...
+		c.coordinator = newCoordinator(c, c.hbstream)
+		c.coordinator.runBgJob()
```
Comment (ShiKaiWi) [inline server/cluster/cluster.go:52]: It seems the `runBgJob` will block the current goroutine but it should not.

File: server/cluster/coordinator.go [hunk 7]
```diff
@@ -5,25 +5,65 @@ package cluster
...
+		cluster: cluster,
+
+		ctx:          ctx,
+		cancel:       cancel,
+		eventHandler: schedule.NewEventHandler(hbstream),
 	}
-	return nil
+}
+
+func (c *coordinator) runBgJob() {
```
Comment (ShiKaiWi) [inline server/cluster/coordinator.go:43]: Provide a concrete name (e.g. dispatchNoneEventPeriodically) if no more jobs are run in background.

File: server/cluster/manager_test.go [hunk 8]
```diff
@@ -6,6 +6,8 @@ import (
 	"context"
 	"testing"
 
+	"github.com/CeresDB/ceresmeta/server/schedule"
+
```
Comment (ShiKaiWi) [inline server/cluster/manager_test.go:10]: These empty lines seems not necessary?

File: server/schedule/event_handler.go [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+func (open *OpenEvent) ToHeartbeatResp() *metaservicepb.NodeHeartbeatResponse {
+	// TODO: add response header
+	return &metaservicepb.NodeHeartbeatResponse{Timestamp: uint64(time.Now().UnixMilli()), Cmd: &metaservicepb.NodeHeartbeatResponse_OpenCmd{
+		OpenCmd: &metaservicepb.OpenCmd{
+			ShardIds: open.ShardIDs,
+		},
+	}}
+}
+
+type EventHandler struct {
```
Comment (ShiKaiWi) [inline server/schedule/event_handler.go:37]: `EventHandler` seems defined for `coordinator` and shall we put to `cluster` package?

File: server/cluster/manager.go [hunk 10]
```diff
@@ -39,11 +40,12 @@ type managerImpl struct {
 	lock     sync.RWMutex
 	clusters map[string]*Cluster
 
-	storage storage.Storage
+	storage  storage.Storage
+	hbstream *schedule.HeartbeatStreams
```
Comment (ShiKaiWi) [inline server/cluster/manager.go:44]:
```suggestion
	hbstreams *schedule.HeartbeatStreams
```

File: server/cluster/cluster.go (hunk 5 above)
Comment (chunshao90) [reply server/cluster/cluster.go:50]: #29  will fix this

File: server/cluster/coordinator.go (hunk 7 above)
Comment (chunshao90) [reply server/cluster/coordinator.go:43]: `dispatchNoneEventPeriodically` is one job, maybe more schedulers run background.

File: server/schedule/event_handler.go (hunk 9 above)
Comment (chunshao90) [reply server/schedule/event_handler.go:37]: HeartBeatStreams need metaservicepb.NodeHeartbeatResponse as input, EventHandler abstract `Event` for cluster dynamic sheduling in ceresmeta.

**Review by ShiKaiWi — APPROVED** [review]

**Pull request merged**